src/io/srt_io.ml

```diff
@@ -0,0 +1,439 @@
+    Srt.bind s bind_address;
+    Srt.listen s 1;
+    self#log#info "Setting up socket to listen at %s"
+      (self#string_of_address bind_address);
+
+  method private create_decoder should_stop socket =
+    let create_decoder =
+      match
+        Decoder.get_stream_decoder format kind
+      with
+        | Some d -> d
+        | None -> raise Harbor.Unknown_codec
+    in
+    let buf = Buffer.create payload_size in
+    let tmp = Bytes.create payload_size in
+    let read len =
+      if Buffer.length buf < len then
+       begin
+        if should_stop () then raise Done;
+        let input = Srt.recvmsg socket tmp payload_size in
+        if input = 0 then raise End_of_file;
+        Buffer.add_subbytes buf tmp 0 input
+       end;
+      let len = min len (Buffer.length buf) in
+      let ret =
+        Buffer.sub buf 0 len
+      in
+      Utils.buffer_drop buf len;
+      ret,len
+    in
+    create_decoder { Decoder.
+      read = read ;
+      tell = None;
+      length = None;
+      lseek = None }
+
+  method private feed (should_stop,has_stopped) =
+    try
+      if should_stop () then raise Done;
+      let client, origin =
+        Srt.accept self#get_socket
+      in 
+      self#log_origin origin;
+      on_connect ();
+      Generator.set_mode generator `Undefined ;
+      let decoder =
+        self#create_decoder should_stop client
+      in
+      while true do
+        if should_stop () then raise Done;
+        let buffered = Generator.length abg in
+        if max_ticks <= buffered then
+          Thread.delay (Frame.seconds_of_audio (buffered-3*max_ticks/4));
+        if should_stop () then raise Done;
+        decoder.Decoder.decode generator
+      done
+    with
+      | e ->
+          (* Feeding has stopped: adding a break here. *)
+          Generator.add_break ~sync:`Drop generator ;
+          on_disconnect ();
+          has_stopped ();
+          self#close_socket;
+          self#log#severe "Feeding stopped: %s." (Printexc.to_string e);
+          if not (should_stop ()) then
+            self#feed  (should_stop,has_stopped)
+    
+  method wake_up act =
+    super#wake_up act ;
+    begin match feeding_thread with
+      | None -> ()
+      | Some (_,wait) -> wait ()
+    end ;
+    feeding_thread <-
+      Some (Tutils.stoppable_thread self#feed self#id)
+  
+  method sleep =
+    begin match feeding_thread with
+      | None -> ()
+      | Some (kill,_) -> kill ()
+    end ;
+    self#close_socket;
+    feeding_thread <- None
+end
+  
+let () =
+  let kind = Lang.univ_t 1 in
+  Lang.add_operator "input.srt"
+    ~kind:(Lang.Unconstrained kind)
+    ~category:Lang.Input
+    ~descr:"Receive a SRT stream."
+    [ "bind_address", Lang.string_t, Some (Lang.string "0.0.0.0"),
+      Some "Address to bind on the local machine.";
+
+      "port", Lang.int_t, Some (Lang.int 8000),
+      Some "Port to bind on the local machine (note: ports in SRT are \
+            not related to TCP ports.)";
```

Comment:
This is confusing, maybe it should be pointed out that these are UDP ports ? Wouldn't it be simplier to have bind_address like 0.0.0.0:8000 ?

Reply:
I'll clarify. Here's what the doc says, which isn't very clear either:
>The term "port" used in SRT is occasionally identical to the term "UDP port". However SRT offers more flexibility than UDP (or TCP, if we think about the more logical similarity) because it manages ports as its own resources. For example, one port may be shared between various services.